### PR TITLE
Prototype of authentication for users

### PR DESCRIPTION
**Pull request by HermanPlay**
To test it, you need .env file with credentials. Text me if you want me to send it to you.

It does not pass tests, since middleware restricts accessing api, if you are not logged in. Kappa

### CODE REVIEW

File: app/api/main.py [hunk 1]
```diff
@@ -33,17 +40,85 @@ async def init(app: FastAPI):
...
+                status_code=exc.status_code,
+                content={"detail": exc.detail},
+            )
+
+
+app.add_middleware(UserValidationMiddleware, excludes=["/login", "/auth", "/logout"])
+app.add_middleware(SessionMiddleware, secret_key="secret-string")
+
+
+config = Config("config.env")  # read config from .env file
```
Comment (matined) [inline app/api/main.py:71]: can it read from env vars instead of .env file? Maybe it already does it?

File: app/api/main.py [hunk 2]
```diff
@@ -33,17 +40,85 @@ async def init(app: FastAPI):
...
+            return response
+        except HTTPException as exc:
+            return JSONResponse(
+                status_code=exc.status_code,
+                content={"detail": exc.detail},
+            )
+
+
+app.add_middleware(UserValidationMiddleware, excludes=["/login", "/auth", "/logout"])
+app.add_middleware(SessionMiddleware, secret_key="secret-string")
```
Comment (matined) [inline app/api/main.py:68]: does the "secret-string" name refer to something?

File: app/api/main.py [hunk 3]
```diff
@@ -33,17 +40,85 @@ async def init(app: FastAPI):
...
 async def root() -> dict:
     """
     Root endpoint. Returns a simple message for testing purposes.
     :return: A "Hello World" message.
     """
     return {"message": "Hello World"}
+
+
+@app.get("/me")
+async def me(user=Depends(get_current_user)):
```
Comment (matined) [inline app/api/main.py:102]: can we move all the endpoints related to authentication and authorization to their own router?

File: app/api/main.py [hunk 4]
```diff
@@ -33,17 +40,85 @@ async def init(app: FastAPI):
     lifespan=init,
 )
 
+
+class UserValidationMiddleware(BaseHTTPMiddleware):
```
Comment (matined) [inline app/api/main.py:44]: let's move this class to some other file

**Review by matined — CHANGES_REQUESTED** [review]

File: app/api/main.py (hunk 1 above)
Comment (HermanPlay) [reply app/api/main.py:71]: We can add those settings to Settings class, which indeed can read from env vars

File: app/api/main.py (hunk 2 above)
Comment (HermanPlay) [reply app/api/main.py:68]: Basically its a key to cypher our sessions, according to official [documentation](https://www.starlette.io/middleware/#sessionmiddleware) this key should be a random string.

**Comment by HermanPlay** [comment]
@matined pls review

**Comment by HermanPlay** [comment]
> review ongoing, please be patient mr.f

Thank you Mr S

**Comment by TheRealSeber** [comment]
please mr.f be patient more. some day mr nędzi will find some time for you.

File: app/api/routers/protected.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from fastapi import APIRouter, Depends
```
Comment (matined) [inline app/api/routers/protected.py:1]: Let's remove this router, protect all other routers instead and we can merge the PR ;))

**Review by matined — CHANGES_REQUESTED** [review]

**Comment by matined** [comment]
Is it something new? Let me know when I can try it out :))

**Comment by HermanPlay** [comment]
> Is it something new? Let me know when I can try it out :))

You can already. Now after logging in you get bearer token which you can use even to access api from terminal. 

In bearer you have user id encoded with expiration time.

It also adds user to database if user doesn't exist.

**Review by matined — APPROVED** [review]

**Pull request merged**